enhancements/ingress/haproxy-router-compression.md

```diff
@@ -0,0 +1,421 @@
+compression type {{ $compressMimeTypes }}
+{{- end }}
+{{- end -}}
+```
+The new function `allMatches` will be added to the `template_helper.go` code:
+```go
+// Compare two white-space delimited strings and return the latter if all its substrings
+// exist in former
+func allMatches(defined string, supplied string) string {
+	log.V(7).Info("allMatches called", "supplied", supplied, "defined", defined)
+	numFound := 0
+	suppliedRange := strings.Fields(supplied)
+	definedRange := strings.Fields(defined)
+	for _, s := range suppliedRange {
+		for _, d := range definedRange {
+            if d == s { // optimize this
+                log.V(7).Info("allMatches validated", "supplied", s)
+                numFound++
+                break
+            }
+        }
+	}
+	if numFound == len(suppliedRange){
+        return supplied
+    }
+	return ""
+}
+```
+
+The proposed change to IngressControllerSpec exposes the HTTP compression configuration variables at the default level:
+```go
+type IngressControllerSpec struct {
+...
+    // httpCompression defines policy for HTTP traffic compression
+    // 
+    // The default value is no compression
+    //
+    // +optional
+    HTTPCompression *HTTPCompressionPolicy `json:"httpCompression,omitempty"`
+}
+...
+// httpCompressionPolicy defines the compression algorithm and MIME types that need to
+// have compression applied.
+//
+// This field is optional, and its absence implies that compression should not be enabled
+// globally in HAProxy.
+//
+// If httpCompressionPolicy exists, compression should be enabled only for the listed
+// MIME types, and only using the designated compression algorithm.
+type HTTPCompressionPolicy struct {
+
+	// compressionAlgorithm defines the desired compression algorithm.
+	//
+	// +required
+	CompressionAlgorithm CompressionAlgorithmType `json:"compressionAlgorithm,omitempty"`
+	
+	// compressionMIMETypes is a list of MIME types that should have compression applied
+	//
+	// +kubebuilder:MinItems=1
+	// +required
+	CompressionMIMETypes []string `json:"compressionMIMETypes,omitempty"`
+}
+
+// CompressionAlgorithmType defines the type of compression algorithm
+// +kubebuilder:validation:Enum=gzip;raw-deflate
+type CompressionAlgorithmType string
+const (
+	gzipCompressionAlgorithm CompressionAlgorithmType = "gzip"
+	rawdeflateCompressionAlgorithm CompressionAlgorithmType = "raw-deflate"
+)
+```
+
+To set up the IngressControllerSpec to use HTTP compression on all routes, the configuration
+might resemble:
+```yaml
+apiVersion: operator.openshift.io/v1
+kind: IngressController
+metadata:
+  name: default
+  namespace: openshift-ingress-operator
+spec:
+  httpCompression:
+    compressionAlgorithm: gzip
+    compressionMIMETypes:
+    - "text/html"
+    - "text/css"
+    - "application/pdf"
+```
+
+### Route-based HTTP Compression via Annotations
+To expose these variables at the route level, the mechanism will be two route annotations that are processed in the
```

Comment:
The annotation based API for HSTS is a historical mistake.  Knowing what we know now, it is a better idea to modify the route API itself to add a new stanza.  Probably even the same stanza you have for the ingress spec.

Reply:
Based on feedback from the team, we will remove route-based HTTP compression options in this release.